### PR TITLE
Improve test/typing/style/doc for MLTuning

### PR DESCRIPTION
**Pull request by jrapin**
## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/functions/ml/mlfunctionlib.py [hunk 1]
```diff
@@ -3,43 +3,52 @@
...
-from .. import utils
-from .. import corefuncs
-
-# EF = tp.TypeVar("EF", bound="ExperimentFunction")
 
 
+# pylint: disable=too-many-instance-attributes,too-many-locals,too-many-arguments
 class MLTuning(ExperimentFunction):
     """Class for generating ML hyperparameter tuning problems.
+    We propose different possible regressors and different dimensionalities.
```
Comment (jrapin) [inline nevergrad/functions/ml/mlfunctionlib.py:20]: @teytaud 
The init docstring needs to go write after the class def, not the init

File: nevergrad/functions/ml/mlfunctionlib.py [hunk 2]
```diff
@@ -3,43 +3,52 @@
...
+    In each case, Nevergrad will optimize the parameters of a scikit learning.
+
+    Parameters
+    ----------
+    regressor: str
```
Comment (jrapin) [inline nevergrad/functions/ml/mlfunctionlib.py:25]: @teytaud  docstrings must follow a particular style to be parsed and well shown in the documentation

File: nevergrad/benchmark/experiments.py [hunk 3]
```diff
@@ -35,26 +35,25 @@
...
 
     # First, a few functions with constraints.
     optims = ["Shiwa", "DE", "DiscreteOnePlusOne", "PortfolioDiscreteOnePlusOne", "CMA", "MetaRecentering",
-              "DoubleFastGADiscreteOnePlusOne", "PSO","BO"]
+              "DoubleFastGADiscreteOnePlusOne", "PSO", "BO"]
     for dimension in [None, 1, 2, 3]:
         for regressor in ["mlp", "decision_tree", "decision_tree_depth"]:
             for dataset in (["boston", "diabetes"] if dimension is None else ["artificialcos", "artificial", "artificialsquare"]):
                 function = MLTuning(regressor=regressor, data_dimension=dimension, dataset=dataset)
-                assert dataset[:10] != "artificial" or dimension is not None
```
Comment (jrapin) [inline nevergrad/benchmark/experiments.py:49]: @teytaud checking that arguments make sense is better done in the init, so that new xp don't need to repeat those lines everytime

File: nevergrad/benchmark/experiments.py [hunk 4]
```diff
@@ -35,26 +35,25 @@
 
 
 @registry.register
-def mltuning(seed: tp.Optional[int] = None, overfitting: bool=False) -> tp.Iterator[Experiment]:
```
Comment (jrapin) [inline nevergrad/benchmark/experiments.py:38]: @teytaud overfitting is not used, is that expected?

File: nevergrad/functions/ml/mlfunctionlib.py [hunk 5]
```diff
@@ -3,43 +3,52 @@
...
-    def _ml_parametrization(self,
-                            depth: int,  # Parameters for regression trees.
-                            criterion: str, 
-                            min_samples_split: float,
-                            solver: str,  # Parameters for neural nets.
-                            activation: str,
-                            alpha: float,
-                            learning_rate: str,
-                            regressor: str,  # Choice of learner.
-                            noise_free: bool  # Whether we work on the test set (the real cost) on an approximation (CV error on train).
```
Comment (jrapin) [inline nevergrad/functions/ml/mlfunctionlib.py:41]: given the definition, maybe `noise_free` is not the right name, it's not actual noise

File: nevergrad/functions/ml/mlfunctionlib.py [hunk 6]
```diff
@@ -101,6 +102,8 @@ def __init__(self, regressor: str, data_dimension: tp.Optional[int] = None, data
         self.add_descriptors(regressor=regressor, data_dimension=data_dimension, dataset=dataset, overfitter=overfitter)
         self.name = regressor + f"Dim{data_dimension}"
         self.num_data: int = 0
+        # Dimension does not make sense if we use a real world dataset.
+        assert bool("artificial" in dataset) == bool(data_dimension is not None)
```
Comment (jrapin) [inline nevergrad/functions/ml/mlfunctionlib.py:106]: check added here instead of in the xps

File: nevergrad/functions/ml/test_mlfunctionlib.py [hunk 7]
```diff
@@ -3,76 +3,67 @@
...
+        dict(depth=5, criterion="mse", min_samples_split=0.001),
+        0.0004427973,
+    ),
+    decision_tree_square=(
+        dict(regressor="decision_tree", data_dimension=3, dataset="artificialsquare"),
+        dict(depth=5, criterion="mse", min_samples_split=0.001),
+        0.001891651,
+    ),
+)
+def test_mltuning_values(cls_params: tp.Dict[str, tp.Any], func_params: tp.Dict[str, tp.Any], expected: float) -> None:
```
Comment (jrapin) [inline nevergrad/functions/ml/test_mlfunctionlib.py:61]: This test is much more generic

File: nevergrad/functions/ml/test_mlfunctionlib.py [hunk 8]
```diff
@@ -3,76 +3,67 @@
...
-              regressor="decision_tree"), 0.004955893558348708)
-
-    # Testing a decision tree on SKLearn's Boston.
-    func.rng.seed(17)
-    func5 = mlfunctionlib.MLTuning(regressor="decision_tree", data_dimension=None, dataset="boston")
-    np.testing.assert_almost_equal(func5(depth=5, criterion="mse", min_samples_split=0.001),
-                                   34.460213262464116)
-
-    # Testing a decision tree on SKLearn's Diabetes.
-    func.rng.seed(17)
```
Comment (jrapin) [inline nevergrad/functions/ml/test_mlfunctionlib.py:60]: seeding `func` and not is not useful here since `func` is not used.

File: nevergrad/benchmark/experiments.py (hunk 4 above)
Comment (teytaud) [reply nevergrad/benchmark/experiments.py:38]: If you update you should see it ? now it's used.

File: nevergrad/benchmark/experiments.py (hunk 3 above)
Comment (teytaud) [reply nevergrad/benchmark/experiments.py:49]: Thx.

File: nevergrad/benchmark/experiments.py (hunk 4 above)
Comment (teytaud) [reply nevergrad/benchmark/experiments.py:38]: naivemltuning uses it. It's the case in which the valid error (used in the optimization of hyperparameters) is also the error used in the evaluation function -- this is for distinguishing algorithms good at optimizing the valid loss from algorithms good at not overfitting.

File: nevergrad/functions/ml/mlfunctionlib.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/functions/ml/mlfunctionlib.py:25]: A big thx and a big apologize...

File: nevergrad/functions/ml/mlfunctionlib.py (hunk 5 above)
Comment (teytaud) [reply nevergrad/functions/ml/mlfunctionlib.py:41]: Yes, you can leave this fix to me for one of these evenings.
Could be "overfitting" or "no_test_error"...

File: nevergrad/functions/ml/test_mlfunctionlib.py (hunk 7 above)
Comment (teytaud) [reply nevergrad/functions/ml/test_mlfunctionlib.py:61]: Definitely true...

File: nevergrad/functions/ml/mlfunctionlib.py (hunk 5 above)
Comment (jrapin) [reply nevergrad/functions/ml/mlfunctionlib.py:41]: `no_cv` or `validation` or `test` (or `train`) would seem more precise, but maybe I dont get it

File: nevergrad/benchmark/experiments.py (hunk 4 above)
Comment (jrapin) [reply nevergrad/benchmark/experiments.py:38]: @teytaud no, I confirm after merging, the parameter is not used in `mltuning`, so `naivemltuning` is just `mltuning` again.
This is something `pylint` can help you catch, maybe we'll set up the commit hook some day ;)

File: nevergrad/functions/ml/mlfunctionlib.py (hunk 5 above)
Comment (teytaud) [reply nevergrad/functions/ml/mlfunctionlib.py:41]: The name "noise_free" was chosen in the sense that the objective function, when it is True, has no randomness, it is deterministic, and evaluation_function == func. In this case,
it is the cross-validation loss.

When it is false, func returns the cross-validation error on the training data, whereas evaluation_function returns the error on separate test data.

**Comment by teytaud** [comment]
Would it make sense to have a directory "experimental" for things which come from the web, not yet widely analyzed, but already shared ? This could help for having possibly expensive simulators, without extensive validation, and possibly without tests run automatically in the CI.

Having them included somewhere helps for discussing with external contributors, and so that anyone can comment.

File: nevergrad/benchmark/experiments.py (hunk 4 above)
Comment (jrapin) [reply nevergrad/benchmark/experiments.py:38]: #669 adds a CI test that all variables are used, it's catching the `overfitting` variable. When you have fixed it I'll merge it and there will be less risk for this kind of issue ;)

**Comment by jrapin** [comment]
> Would it make sense to have a directory "experimental" for things which come from the web, not yet widely analyzed, but already shared ? This could help for having possibly expensive simulators, without extensive validation, and possibly without tests run automatically in the CI.
> 
> Having them included somewhere helps for discussing with external contributors, and so that anyone can comment.

If you have ideas in mind why not, but whenever possible I prefer it as an actual benchmark since this way it's robust.
